### ISSUE TITLE
Extended variable can be assigned, but literal object cannot be assigned.

### ISSUE THREAD

**Issue by MartinYounghoonKim** [microsoft/TypeScript#36122]
**Search Terms:**
2 Days

**Code**

```ts
interface A {
    a: number;
    b: string;
}
interface B extends A {
    c: string;
}
const test = {
    a: 10,
    b: "foo",
    c: "bar"
}
const a: A[] = [test];
const b: A[] = [{
    a: 10,
    b: "foo",
    c: "bar" // ❌ not assignable type error
}];
```

I don't understand what difference `a` and `b` variables. And I got a similar error.

```ts
interface A {
    a: number;
    b: string;
}
interface B extends A {
    c: string;
}
const test = {
    a: 10,
    b: "foo",
    c: "bar"
}
const a: A[] = [test, {
    a: 10,
    b: "foo",
    c: "bar"
}];
```

It seem to be able to assigned. But when I try to add the wrong element to the array, I see an error that couldn't be displayed.

```ts
const a: A[] = [test, {
    a: 10,
    b: "foo",
    c: "bar"  // ❌ not assignable type error, why this error displayed at this time ?
}, 1];  // ❌ this type error is understandable
```

**Expected behavior:**
All variables passed without error or all the same error occurs.

**Actual behavior:**
All variables are same , but they return another result.

**Playground Link:** 
- [Playground Link](https://www.typescriptlang.org/play/?ssl=22&ssc=13&pln=23&pc=13#code/JYOwLgpgTgZghgYwgAgILIN4FgBQz-JwBcyIArgLYBG0A3LgclSQM5hSgDm9OAvrqEixEKAELIIAD0ggAJizSYGBBK3Zce-HAgD2INskgGAvErwFiyAIwAGADTL8zZACIYOnS4fn8q11TgoF1wtXX0wQhJUAG0AXWRTaKMwWJ4wg2cY+MTkuzNGS1tvRmc3Dy9HZD8XAKCQ1Nx0iL8shOQkiDY87B9I63tK0vdPYpUSGsCXfAB6aeRAGXJkELyrVJm5xZAdCLgWFmBOEDgqABsUMABPAAcUaCgdKEa9A1kouLbonoKSIsHx4YqvWqtSmswWpG2hD2ByOp3O11uUHujz4qSAA)

**Comment by IllusionMH** [comment]
This one functions as designed. Object literals shouldn't provide properties that type expects.
`A` only has `a` and `b`, therefore passing 1 more property triggers this error.

Simplified example:
```ts
const a: { a: number } = { a: 1, b: 'extra' };
```

For more details and workaround see [Interfaces > Excess Property Checks](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks)

**Comment by MartinYounghoonKim** [comment]
@IllusionMH 
thank you for reply. According to what you said, shouldn't this be?

```ts
interface A {
    a: number;
    b: string;
}
interface B extends A {
    c: string;
}
const test = {
    a: 10,
    b: "foo",
    c: "bar"
}
const a: A[] = [test]; 
```

**Comment by IllusionMH** [comment]
Excess property checking is only applied to object literals which are usually typed by developers.
`const a: A[] = [test];` uses variable (but it might as well be a function argument) so no excess property checks are applied.

On the other hand these will be errors
```ts
const test: A = {
    a: 10,
    b: "foo",
    c: "bar"
}
```
or
```ts
function foo(a: A) {}

foo({
    a: 10,
    b: "foo",
    c: "bar"
});
```

**Comment by MartinYounghoonKim** [comment]
@IllusionMH 
thanks for reply. sorry, I trying to edit my question.
but what about second example ?

**Comment by IllusionMH** [comment]
Excess property checks are performed only on **fresh object literal** types.

When you assign array to variable, type of array value will be compared to `A`.

In first case there will be union of `{a: number; b: string; c: string;} | {a: number; b: string; c: string;}(fresh)` and during union types deduplication only non-fresh type will be left `{a: number; b: string; c: string;}`. Since TS will compare non-fresh object to `A` there will be no type errors, so no need to check further.

In second case - type will be `number | {a: number; b: string; c: string;}` after union deduplication. Because this union is not assignable to `A` TS will check every item of array literal independently to report problems. When compared by item fresh object literal will be checked for excess properties and error will be reported.

So to sum up: no error in first cases because `test` and object have same structure there fore freshness will be lost (widened I suppose), no assignability problems without freshness, no reason to compare each array element against target type.

**Comment by MartinYounghoonKim** [comment]
@IllusionMH 
I'm appreciate your reply. I close this issue.

**Comment by FrankChencc** [comment]
See [Type Inference](https://www.typescriptlang.org/docs/handbook/type-inference.html)

In first case ，test's inferred type is` {a: number; b: string; c: string; }` when a fresh object literal is assigned to it. 
```
//  test's inferred type is` {a: number; b: string; c: string; }
const test = {
    a: 10,
    b: "foo",
    c: "bar"
}
```
Then contextual typing applies in  array literals，the best common type is ` {a: number; b: string; c: string; }`,  and  for each property of `A`, it can be find a corresponding compatible property in the best common type, so the array can be assigned to` A[]`, and the type of this object literal is widened. see this [pull request](https://github.com/microsoft/TypeScript/pull/3823)
```
//  the best common type is ` {a: number; b: string; c: string; }` 
[test, {
    a: 10,
    b: "foo",
    c: "bar" 
}];
```
